### PR TITLE
affiliate-server with queue for amz

### LINKED ISSUE #4220: Open Library Amazon API usage exceeding limit
Dup of #3967

Received email from Amazon API:

> Hello,
> 
> Your Associates account is at risk of closure.
> 
> Why?
> You are trying to submit requests that exceed your account’s usage limit. You must not exceed, or if you build and release an application that calls PA API, each copy of that application that is installed by an end user will not exceed, any limit on calls per second set forth in any Specifications.
> 
> What’s next?
> Within five business days please correct the violations and notify us when you are in compliance through our customer service contact page: https://affiliate-program.amazon.com/home/contact. Please choose the subject 'Warning/Information Request Response' from the dropdown menu, and be sure to reference Issue Code 83441-US in the comments field.
> 
> If not done within five business days, we will close your account.

## Stakeholders
@mekarpeles @bfalling

### PR DESCRIPTION
**Pull request by mekarpeles**
Closes #4220,#3211
Relates to  #4180 #4150 

### Technical

Stand-alone affiliate-server --
1. accepts submit requests from all/any workers.
2. workers receive response of submission (as well as any cached hits)
3. worker may have to submit multiple requests (waiting  queue to hit threshold length) before their result in cache
4. affiliate-server maintains a queue of `threshold`. When this threshold is passed, bulk amazon submit request made, product results added to global affiliate-server cache, queue is cleared, and final results are returned.

XXX likely critical section / race conditions. Should be handled by an actual queue.

### Testing

If you load each of these in a browser, they should return almost immediately, until we hit the last request (which will trigger the amazon call). If we then reload the previous 4 requests, each should reply almost immediately from cache:
http://dev.openlibrary.org:31337/isbn/1111111111
http://dev.openlibrary.org:31337/isbn/3836582538
http://dev.openlibrary.org:31337/isbn/1400033438
http://dev.openlibrary.org:31337/isbn/0425232204
http://dev.openlibrary.org:31337/isbn/0394810112

### CODE REVIEW

File: scripts/affiliate-server [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        web.amazon_api = AmazonAPI(
+            web.ol_config.amazon_api.get('key'),
+            web.ol_config.amazon_api.get('secret'),
+            web.ol_config.amazon_api.get('id'),
+            throttling=0.5
+        )
+        web.amazon_cache = {}
+        web.amazon_queue = set()
+    except AttributeError as e:
+        print(e)
```
Comment (cclauss) [inline scripts/affiliate-server:112]:
```suggestion
        logger.exception("Initializing AmazonAPI")
        print(e)
```

File: scripts/affiliate-server [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                }
+            try:
+                queue = list(web.amazon_queue)
+                web.amazon_queue = set()
+                products = web.amazon_api.get_products(queue, serialize=True)
+                for product in products:
+                    web.amazon_cache[product['isbn_10'][0]] = product
+                    hits[product['isbn_10'][0]] = product
+            except Exception:
+                logger.exception("amazon_api.get_product({}, serialize=True)".format(web.amazon_queue))
```
Comment (cclauss) [inline scripts/affiliate-server:85]:
```suggestion
                logger.exception("amazon_api.get_product({}, serialize=True)".format(queue))
```

File: scripts/affiliate-server (hunk 1 above)
Comment (cclauss) [reply scripts/affiliate-server:112]: Can `config.get()` raise an AttributeError?

**Review by cclauss — CHANGES_REQUESTED** [review]

**Comment by cclauss** [comment]
Will this require a new Sentry project?

File: scripts/affiliate-server (hunk 1 above)
Comment (mekarpeles) [reply scripts/affiliate-server]: checking if all args exist instead

File: scripts/affiliate-server [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+            #except Exception:
+            #    logger.exception("amazon_api.get_product({}, serialize=True)".format(queue))                
+            #    web.amazon_queue.update(queue)  # Careful, request may enter fail-loop from bad IDs
+
+        if cache.memcache_cache.get('amazon_product_%s' % isbn13):
+            return {
+                "status": "success",
+                "hit": cache.memcache_cache.get('amazon_product_%s' % isbn13)
+            }
+        return {
```
Comment (mekarpeles) [inline scripts/affiliate-server:94]: XXX blocker: All return types need to be application/json and json.dumps. Currently, it's returning a python object, not valid json

Comment (cclauss) [reply scripts/affiliate-server:94]: `status` is a str, `queue` is an int, `threshold` is an int.

File: scripts/affiliate-server [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    web.amazon_api = None
+    args = [
+        web.ol_config.amazon_api.get('key'),
+        web.ol_config.amazon_api.get('secret'),
+        web.ol_config.amazon_api.get('id')
+    ]
+    if all(args):
+        web.amazon_api = AmazonAPI(*args, throttling=0.5)
+        web.amazon_queue = set()
+        logger.exception("Initializing AmazonAPI")
```
Comment (cclauss) [inline scripts/affiliate-server:119]:
```suggestion
        logger.info("Initializing AmazonAPI")
```
[`logger.exception()`](https://docs.python.org/3/library/logging.html#logging.Logger.exception) _should only be called from an exception handler_.

Do we want to deal with the else case?

File: openlibrary/plugins/openlibrary/api.py [hunk 5]
```diff
@@ -66,6 +66,11 @@ def GET(self):
         sorts = i.sorts.split(',')
         page = int(i.page)
         limit = int(i.limit)
+
+        print("query: " + i.q)
+        print("subject: " + i.subject)
+        print("work_id: " + i.work_id)
```
Comment (cclauss) [inline openlibrary/plugins/openlibrary/api.py:72]: Should these remain?  Should they be `logger.info()` instead?

File: scripts/affiliate-server [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            return json.dumps({
+                "error": "rejected_isbn",
+                "isbn10": isbn10,
+                "isbn13": isbn13
+            })
+
+        # Cache lookup by isbn13. If there's a miss, add to queue
+        if not cache.memcache_cache.get('amazon_product_%s' % isbn13):
+            # Misses (which will get submitted to amz as ASIN) are isbn10
+            # We should track when (queue % 20) occurs
```
Comment (cclauss) [inline scripts/affiliate-server:75]: 20?

File: scripts/affiliate-server [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+    web.amazon_api = None
+    args = [
+        web.ol_config.amazon_api.get('key'),
+        web.ol_config.amazon_api.get('secret'),
+        web.ol_config.amazon_api.get('id')
+    ]
+    if all(args):
+        web.amazon_api = AmazonAPI(*args, throttling=0.9)
+        web.amazon_queue = set()
```
Comment (cclauss) [inline scripts/affiliate-server:128]: My sense is that we should _really_ use a queue... https://docs.python.org/3/library/queue.html

Python sets are guaranteed to be unordered.  Python queues are FIFO which is I think what we want.

At the top of the file: `from queue import Queue`
```suggestion
        web.amazon_queue = Queue()
```
If the value is in Memcached then return it
else `web.amazon_queue.put_nowait(value)` and return a promise

... acquision logic ...
batch = set()  # removes any dups...
while not amazon_queue.empty() and len(batch) < threshold:
    batch.add(amazon_queue.get())
if batch:
    amazon_lookup(batch)

**Comment by cclauss** [comment]
```
import time
from queue import Queue
from threading import Thread

API_THROTTLE_SECONDS = 0.9
API_MAX_ITEMS_PER_CALL = 10
API_MAX_WAIT_SECONDS = 5

amazon_queue = Queue()
start_time = time.time()


def wait_seconds_remaining():
    return API_MAX_WAIT_SECONDS - (start_time - time.time())


def amazon_lookup():
    global start_time
    while True:
        time.sleep(API_THROTTLE_SECONDS)
        if amazon_queue.empty:
            continue
        batch = set()  # no duplicates
        while len(batch) < API_MAX_ITEMS_PER_CALL and wait_seconds_remaining() > 0:
            batch.add(amazon_queue.get(), timeout=wait_seconds_remaining())
        if batch:
            process_batch(list(batch))


Thread(amazon_lookup).start()
```

**Comment by cclauss** [comment]
Am I doing something wrong with the ports?  When I launch on localhost, I get 
```
Creating openlibrary_covers_1           ... done
Creating openlibrary_affiliate-server_1 ... done
Creating openlibrary_db_1               ... done
Creating openlibrary_home_1             ... done
Creating openlibrary_memcached_1        ... done
Creating openlibrary_solr_1             ... done
Creating openlibrary_infobase_1         ... done
Creating openlibrary_web_1              ... done
Attaching to openlibrary_web_1, openlibrary_infobase_1, openlibrary_db_1, openlibrary_covers_1, openlibrary_home_1, openlibrary_affiliate-server_1, openlibrary_memcached_1, openlibrary_solr_1
affiliate-server_1  | Python 3.8.6
affiliate-server_1  | /openlibrary/docker/ol-affiliate-server-start.sh: scripts/affiliate-server: /usr/bin/env: bad interpreter: Permission denied
openlibrary_affiliate-server_1 exited with code 126
```

**Comment by cclauss** [comment]
@mekarpeles @cdrini Please review carefully.  Who do I test this given that it only runs on prod?

File: openlibrary/plugins/openlibrary/api.py (hunk 5 above)
Comment (mekarpeles) [reply openlibrary/plugins/openlibrary/api.py:72]: Nope!

File: scripts/affiliate-server (hunk 6 above)
Comment (mekarpeles) [reply scripts/affiliate-server]: Sorry -- this was an arbitrary attempt at achieving the same thing as sampling statsd (e.g. every time the queue passes some multiple of 20)

**Comment by mekarpeles** [comment]
@cclauss re: testing, I think the right strategy is for us to try to set up / test this on ol-home0 (using docker-compose.production.yml)

**Review by cclauss — APPROVED** [review]

**Comment by cclauss** [comment]
Close this?

**Review by cdrini — APPROVED** [review]
Lgtm; we tested:

- amazon requests made
- requests stored in memcached and not in memory
- /isbn/ import works on staging

**Pull request merged**